batch/batch/cloud/azure/driver/create_instance.py

```diff
@@ -0,0 +1,471 @@
+from typing import Any, Dict
+import base64
+import json
+import logging
+import os
+from shlex import quote as shq
+
+from gear.cloud_config import get_global_config
+
+from ....batch_configuration import (DOCKER_ROOT_IMAGE, DOCKER_PREFIX, DEFAULT_NAMESPACE,
+                                     INTERNAL_GATEWAY_IP)
+from ....file_store import FileStore
+from ....instance_config import InstanceConfig
+
+from ...resource_utils import unreserved_worker_data_disk_size_gib
+from ..resource_utils import azure_machine_type_to_worker_type_and_cores
+
+
+log = logging.getLogger('create_instance')
+
+BATCH_WORKER_IMAGE = os.environ['HAIL_BATCH_WORKER_IMAGE']
+
+log.info(f'BATCH_WORKER_IMAGE {BATCH_WORKER_IMAGE}')
+
+
+def create_vm_config(
+        file_store: FileStore,
+        resource_rates: Dict[str, float],
+        location: str,
+        machine_name: str,
+        machine_type: str,
+        activation_token: str,
+        max_idle_time_msecs: int,
+        local_ssd_data_disk: bool,
+        data_disk_size_gb: int,
+        preemptible: bool,
+        job_private: bool,
+        subscription_id: str,
+        resource_group: str,
+        ssh_public_key: str,
+        instance_config: InstanceConfig,
+) -> dict:
+    _, cores = azure_machine_type_to_worker_type_and_cores(machine_type)
+
+    if job_private:
+        unreserved_disk_storage_gb = data_disk_size_gb
+    else:
+        unreserved_disk_storage_gb = unreserved_worker_data_disk_size_gib(data_disk_size_gb, cores)
+    assert unreserved_disk_storage_gb >= 0
+
+    worker_data_disk_name = 'data-disk'
+
+    if local_ssd_data_disk:
+        data_disks = []
+        disk_location = '/dev/disk/azure/resource'
+    else:
+        data_disks = [
+            {
+                "name": "[concat(parameters('vmName'), '-data')]",
+                "lun": 2,  # because this is 2, the data disk will always be at 'sdc'
+                "managedDisk": {
+                    "storageAccountType": "Standard_LRS"
+                },
+                "createOption": "Empty",
+                "diskSizeGB": data_disk_size_gb,
+                "deleteOption": 'Delete'
+            }
+        ]
+        disk_location = '/dev/disk/azure/scsi1/lun2'
+
+    make_global_config = ['mkdir /global-config']
+    global_config = get_global_config()
+    for name, value in global_config.items():
+        make_global_config.append(f'echo -n {shq(value)} > /global-config/{name}')
+    make_global_config_str = '\n'.join(make_global_config)
+
+    assert instance_config.is_valid_configuration(resource_rates.keys())
+
+    startup_script = r'''#cloud-config
```

Comment:
admittedly somewhat confusing that Azure calls this a startup *script* when its a configuration file. 🤷